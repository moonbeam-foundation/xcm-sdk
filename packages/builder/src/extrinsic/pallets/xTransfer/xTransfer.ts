import { ExtrinsicConfig } from '../../../types/substrate/ExtrinsicConfig';
import type { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { getExtrinsicAccount } from '../../ExtrinsicBuilder.utils';

const pallet = 'xTransfer';

export function xTransfer() {
  return {
    transfer: () => {
      const method = 'transfer';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: ({ destinationAddress: address, asset, destination }) =>
            new ExtrinsicConfig({
              module: pallet,
              func: method,
              getArgs: () => [
                {
                  id: {
                    Concrete: {
                      parents: 0,
                      interior: 'Here',
                    },
                  },
                  fun: {
                    Fungible: asset.amount,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      getExtrinsicAccount(address),
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: ({ destinationAddress, asset, destination }) =>
            new ExtrinsicConfig({
              module: pallet,
              func: method,
              getArgs: () => [
                {
                  id: {
                    Concrete: {
                      parents: 1,
                      interior: {
                        X2: [
                          {
                            Parachain: destination.parachainId,
                          },
                          {
                            PalletInstance: asset.getAssetPalletInstance(),
                          },
                        ],
                      },
                    },
                  },
                  fun: {
                    Fungible: asset.amount || 1n,
                  },
                },
                {
                  parents: 1,
                  interior: {
                    X2: [
                      {
                        Parachain: destination.parachainId,
                      },
                      getExtrinsicAccount(destinationAddress),
                    ],
                  },
                },
                {
                  refTime: 5_000_000_000,
                  proofSize: 0,
                },
              ],
            }),
        }),
      };
    },
  };
}
