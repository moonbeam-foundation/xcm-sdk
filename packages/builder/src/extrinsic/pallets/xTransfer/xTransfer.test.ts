import { describe, expect, it } from 'vitest';

import {
  buildParachainParamsMock,
  buildParamsMock,
} from '../../../../fixtures';
import { xTransfer } from './xTransfer';

describe('xTransfer', () => {
  describe('transfer', () => {
    describe('here', () => {
      const extrinsic = xTransfer().transfer().here().build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      const extrinsic = xTransfer().transfer().X2().build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('parachain', () => {
      const extrinsic = xTransfer()
        .transfer()
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
