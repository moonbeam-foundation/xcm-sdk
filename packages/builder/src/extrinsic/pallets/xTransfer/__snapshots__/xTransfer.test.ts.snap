// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`xTransfer > transfer > here > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTransfer",
}
`;

exports[`xTransfer > transfer > here > should get correct arguments 1`] = `
[
  {
    "fun": {
      "Fungible": 99000000000n,
    },
    "id": {
      "Concrete": {
        "interior": "Here",
        "parents": 0,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;

exports[`xTransfer > transfer > parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTransfer",
}
`;

exports[`xTransfer > transfer > parachain > should get correct arguments 1`] = `
[
  {
    "fun": {
      "Fungible": 99000000000n,
    },
    "id": {
      "Concrete": {
        "interior": {
          "X2": [
            {
              "Parachain": 2032,
            },
            {
              "PalletInstance": 10,
            },
          ],
        },
        "parents": 1,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 2032,
        },
        {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;

exports[`xTransfer > transfer > x2 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transfer",
  "getArgs": [Function],
  "module": "xTransfer",
}
`;

exports[`xTransfer > transfer > x2 > should get correct arguments 1`] = `
[
  {
    "fun": {
      "Fungible": 99000000000n,
    },
    "id": {
      "Concrete": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "PalletInstance": 10,
            },
          ],
        },
        "parents": 1,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;
